### ISSUE TITLE
[Question] AcquireTokenSilent behaviour for updated claims

### ISSUE THREAD

**Issue by IhnatKlimchuk** [AzureAD/microsoft-authentication-library-for-dotnet#1707]
**Which Version of MSAL are you using ?**
MSAL 4.9.0

**Platform**
xamarin android

**What authentication flow has the issue?**
* Desktop / Mobile
    * [x] Interactive/Silent
    * [ ] Integrated Windows Auth
    * [ ] Username Password
    * [ ] Device code flow (browserless)
* Web App
    * [ ] Authorization code
    * [ ] OBO
* Web API
    * [ ] OBO

**Is this a new or existing app?**
This is a new app or experiment

**Repro**

```csharp
...
//1. Setup:
_publicClientApplication = PublicClientApplicationBuilder
                .Create(clientId)
                .WithB2CAuthority("YetAnotherAuthority")
                .WithRedirectUri($"msal{clientId}://auth")
                .Build();
...
//2. First SignIn
AuthenticationResult authResult = await _publicClientApplication
                .AcquireTokenInteractive(someAmazingScopes)
                .WithPrompt(Prompt.SelectAccount)
                .WithParentActivityOrWindow(App.AndroidMainActivity)
                .ExecuteAsync();
...
//3. Second SignIn after updates
AuthenticationResult authResult = await _publicClientApplication
                .AcquireTokenSilent(someAmazingScopes, account)
                .WithForceRefresh(true)
                .ExecuteAsync();
...
```
**Description**
1. I have an android xamarin application. After initial signin I've received all required claims.
2. I do some changes in Azure B2C per user: update "username", add new custom (external) claim "extension_AccountType".
3. I restart application and perform silent signin.

**Expected behavior**
I can see updated "username" and new claim "extension_AccountType" in token.

**Actual behavior**
Token is the same. "exp" and "iat" are updated.

Please correct me if I'm wrong and it's not expected behaviour. What are other possible ways to get/update claims on user side? Can user trigger claims exchange?

Thank you in advance.

**Comment by bgavrilMS** [comment]
Very interesting issue. 

The token cache that MSAL maintains holds on to access tokens, refresh tokens etc. `AcquireTokenSilent` tries to find you an access token from the cache first, without going to AAD. This works for the lifetime of the access token (1h). So during this interval you will not see claim changes, because you are just getting cached results. 

Workarounds: 

1. Wait 1h ... then MSAL will use the refresh_token grant to get you new access and id tokens. These should have your changes
2. Use `WithForceRefresh(true)` in AcquireTokenSilent once, to bypass acquiring the access token. Do no use this everytime, it will slow down your app and, AAD will throttle you if you hit it too often.

**Comment by IhnatKlimchuk** [comment]
@bgavrilMS 

After almost 1 day later and with `WithForceRefresh(true)` I still see old values after `AcquireTokenSilent`. Any ideas why is it happening? I use default `Sign up and sign in` (v1) flow.

**Comment by IhnatKlimchuk** [comment]
@bgavrilMS tested with `Sign up and sign in v2` - behaviour is the same, no update. Can you confirm that claims should be refreshed?

**Comment by bgavrilMS** [comment]
Yes, claims should be refreshed as far as I know. A new id token is issued by the refresh_token flow. 

Can you try to skip the cache altogether (i.e. don't call AcquireTokenSilent) and just use AcquireTokenInteractive ? Are there claims present then?

**Comment by IhnatKlimchuk** [comment]
My steps:
0. In Azure ad b2c UI I update user "name" to "2"
1. I start an app with prevoiusly signed user, that had claim "name" equal "1"
2. In App constructor I do `AcquireTokenSilent` and receive success with claim "name" equal to "1"
3. Call `AcquireTokenInteractive `, enter credentials and receive success with claim "name" equal to "2"
4. After app restart `AcquireTokenSilent` returns success with claim "name" equal to "2"

Should I provide any additional info?

**Comment by IhnatKlimchuk** [comment]
@bgavrilMS Any updates? @henrik-me I haven't received answer, can you please remove label?

**Comment by bgavrilMS** [comment]
Thanks for following up @IhnatKlimchuk - this may well be a B2C bug. I'll investigate and follow up.

**Comment by IhnatKlimchuk** [comment]
@bgavrilMS any updates? Can I help you?

**Comment by bgavrilMS** [comment]
Sorry @IhnatKlimchuk , did not make any progress on this. In your repro steps above, at step 2, can you add a `.WithForceRefresh(true)` so that we're sure the cache is bypassed please? 

I think the issue is not related to the SDK, as we're getting tokens, but to the service. @JasSuri - can you help or add somebody who knows more about this?

**Comment by JasSuri** [comment]
AAD B2C does not update the claim set when doing a silent login (refresh token flow). Therefore the behaviour you're seeing is expected. If you have the user go through the "profile edit" journey, you can obtain the updated claims based on their updates. But if you modify their data on the backend with Graph API, the new claims will not be reflected in the users tokens until a fresh authentication occurs. For now, this is "by design", but of course we are actively working on improving that such that a silent token call will give the latest claim set.

**Comment by IhnatKlimchuk** [comment]
@JasSuri Thank you for answer. Where I can ask/vote for this feature?

**Comment by JasSuri** [comment]
It's being built already, but you can vote for it or create an idea at: https://feedback.azure.com/forums/169401-azure-active-directory?category_id=160596
